module3/lessons/refactoring_api_consumption.md

```diff
@@ -0,0 +1,263 @@
+
+---
+
+## Refactoring Our Controller
+
+So far, most of the code we've written in Rails has fit nicely into the MVC structure, and we haven't explored writing files much beyond controllers, models, and serializers. But now with API consumption, we're handling logic that doesn't fit squarely into any of these components. 
+
+**Brainstorm:** If you were building a vanilla Ruby application, what types of files do you think you would make to extract this logic out of the controller?
+
+<section class="dropdown">
+### Possible Design Pattern #1: Gateways
+
+[Gateway classes](https://mattbrictson.com/blog/gateway-pattern) are very common across many frameworks. They are also referred to as a *service* but this is a very overloaded term in Rails and software development as a whole. This section will refer to these classes as *gateways* but you're welcome to name them however you'd prefer.
+
+Gateways typically encapsulate the logic used for interacting with an external API. Let's refactor our controller to use a gateway. 
+
+*app/controllers/images_controller.rb*
+```ruby
+class Api::V1::ImagesController < ApplicationController
+  def show
+    artist = params[:artist]
+
+    first_photo = ImageGateway.get_first_image(artist)
+    
+    formatted_json = {
+      id: nil,
+      type: "image",
+      attributes: {
+        image_url: first_photo[:url],
+        photographer: first_photo[:photographer],
+        photographer_url: first_photo[:photographer_url],
+        alt_text: first_photo[:alt]
+      }
+    }
+
+    render json: { data: formatted_json }
+  end
+end
+```
+
+Now our Gateway class might look like this:
+```ruby
+class ImageGateway
+  def self.get_first_image(query_term)
+    conn = Faraday.new(url: "https://api.pexels.com") do |faraday|
+      faraday.headers["Authorization"] = Rails.application.credentials.pexels[:key]
+    end
+
+    response = conn.get("/v1/search", { query: query_term })
+    # OR response = conn.get("/v1/search?query=#{artist})
+
+    json = JSON.parse(response.body, symbolize_names: true)
+    json[:photos][0]
+  end
+end
+```
+
+Notice how much logic that extracted out of the controller! Hooray! 
+
+If we are making calls to more than one Pexels endpoint in our application, we could also consider adding additional class methods to our Gateway and extacting the Faraday connection set up into a helper method that can be called in each method.
+
+```ruby
+class ImageGateway
+  def self.get_first_image(query_term)
+    response = connection.get("/v1/search", { query: query_term })
+
+    json = JSON.parse(response.body, symbolize_names: true)
+    json[:photos][0]
+  end
+
+  private
+  
+  def self.connection
```

Comment:
Above this method was called `self.conn` and now it's `self.connection`. I think we should choose one and stick to it for consistency. `def conn` is always more fun. haha

Reply:
oops that's a typo from me! Thanks for catching!